pytorch_to_returnn/converter/converter.py

```diff
@@ -156,6 +157,24 @@ def _make_tf_feed_dict(self, input: Data):
     else:
       for i, size in input.size_placeholder.items():
         d[size] = [self._inputs_np.shape[input.get_batch_axis(i)]] * n_batch  # not so relevant
+    if out_entry:
+      def _get_non_deterministic_tensor_entries(entry_: TensorEntry):
+        """
+        Go recursively through all tensor entries to find non deterministic ones
+        """
+        if not all(call_.module.module.is_deterministic for call_ in entry_.output_from_calls):
```

Comment:
I think you could also stop doing the recursive search when you see some entry with `validated_to_torch` because that means that this should already be part of the feed dict.

Reply:
I don't think so. E.g. in `test_batch_norm`, the `feed_dict` so far only contained the data and the data size placeholder. However, `entry_.validated_to_torch_tf_feed_dict` for the batch norm's output entry contains the batch norm output placeholder. So this is not part of the feed dict yet and it is deterministic, so we should not skip this part in the `session.run` for the check.

Reply:
But when this is deterministic and we have the output value, then whatever comes before this doesn't matter?

Reply:
By "is deterministic", I mean that the module is not random. If we skip whatever comes before, we don't verify that part against the torch ref. So e.g. the batch norm computation would not be part of the check while I think it should be part.

Reply:
I don't exactly understand. This here is just for creating the feed dict, not about the check. Also `_make_tf_feed_dict` is just used for the whole network check. And given some output entry, here we search for all needed feed inputs to calculate that output entry. When you feed in some intermediate entry (because it is random), everything which comes before that is not relevant to calculate the output.

Reply:
> When you feed in some intermediate entry (because it is random), everything which comes before that is not relevant to calculate the output.

Exactly. And I'm arguing that `entry_.validated_to_torch_tf_feed_dict` also contains tensors which are output of regular, deterministic modules. If we would add all of them, they would not be considered in the check.

Reply:
But I did not say to add all. I said the opposite. I would stop the recursive search for this case and not continue adding things.

Reply:
Isn't it possible that there is some random module before that?

Reply:
Why does it matter? `out_entry` depends on this module, so when you feed this, whatever comes before does not matter.

Reply:
I thought it would be possible to have something like input -> some random module -> some deterministic module -> output, where `entry_.validated_to_torch_tf_feed_dict` is set for the output of the deterministic module. Then, we would still need to go deeper because we need to add the random module's output to the `feed_dict`. If you say this case cannot occur, I'll add the suggested stop condition.

Reply:
Ah sorry, my initial comment confused this function with some other case. I thought this is for intermediate checking, not for the whole network checking. For the whole network checking, yes, you are right, nothing else is needed to be added here anymore.